### PR TITLE
nmcli: Disallow Wi-Fi options not supported by nmcli

### PR DESCRIPTION
**Pull request by hummeltech**
##### SUMMARY

Since both the `wifi` & `wifi_sec` options do not have a static list of suboptions (it depends on `nmcli's` version), we should query `nmcli` to determine which options are available on the client. Currently, when applying changes containing invalid/unsupported options, only the first option will be shown, show all of them instead. Also added new option to allow the play to display a warning and continue when invalid/unsupported options are present, the default is to fail.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`net_tools/nmcli.py`

##### ADDITIONAL INFORMATION

When specifying invalid/unsupported Wi-Fi options in the `wifi` and/or `wifi_sec` dict.
### I.E.
* **`ssid` was specified via `wifi.ssid`:**
  ```YAML
  type: wifi
  ssid: SSID_SHOULD_BE_SPECIFIED_HERE
  wifi:
    ssid: SSID_SHOULD_NOT_BE_SPECIFIED_HERE
  ```
  * Before:
    * No error message is given, so the user is unaware of any potential issue/disparity.
  * After:
    ```
    [WARNING]: Ignoring option 'wifi.ssid', it must be specified with option 'ssid'
    ```

* **Invalid option was specified via `wifi`:**
  ```YAML
  type: wifi
  ssid: SSID
  wifi:
    INVALID_1: This is a bad option
    INVALID_2: This is a bad option
    INVALID_3: This is a bad option
  ```
  * Before (note only the first invalid/unsupported option is detected):
    ```
    fatal: [router]: FAILED! => {"changed": false, "msg": "Error: invalid property 'INVALID_1': 'INVALID_1' not among [ssid, mode, band, channel, bssid, rate, tx-power, mac-address, cloned-mac-address, generate-mac-address-mask, mac-address-blacklist, mac-address-randomization, mtu, seen-bssids, hidden, powersave, wake-on-wlan, ap-isolation].\n", "name": "WLAN", "rc": 2}
    ```
  * After:
    ```
    fatal: [router]: FAILED! => {"changed": false, "msg": "Invalid or unsupported option(s): \"wifi.INVALID_1\", \"wifi.INVALID_2\", \"wifi.INVALID_3\""}

    ```

* **Invalid option was specified via `wifi` with `ignore_unsupported_suboptions` enabled:**
  ```YAML
  ignore_unsupported_suboptions: yes
  type: wifi
  ssid: SSID
  wifi:
    INVALID_1: This is a bad option
    INVALID_2: This is a bad option
    INVALID_3: This is a bad option
  ```
  * Before (note only the first invalid/unsupported option is detected):
    ```
    fatal: [router]: FAILED! => {"changed": false, "msg": "Error: invalid property 'INVALID_1': 'INVALID_1' not among [ssid, mode, band, channel, bssid, rate, tx-power, mac-address, cloned-mac-address, generate-mac-address-mask, mac-address-blacklist, mac-address-randomization, mtu, seen-bssids, hidden, powersave, wake-on-wlan, ap-isolation].\n", "name": "WLAN", "rc": 2}
    ```
  * After:
    ```
    [WARNING]: Invalid or unsupported option(s): 'wifi.INVALID_1', 'wifi.INVALID_2', 'wifi.INVALID_3'
    ```

### CODE REVIEW

**Comment by felixfontein** [comment]
Since this contains #3160, let's keep this WIP until that's merged and this PR has been rebased.

File: changelogs/fragments/3141-disallow-options-unsupported-by-nmcli.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - nmcli - query `nmcli` directly to determine available WiFi options
```
Comment (felixfontein) [inline changelogs/fragments/3141-disallow-options-unsupported-by-nmcli.yml:2]:
```suggestion
  - nmcli - query ``nmcli`` directly to determine available WiFi options
```

**Comment by felixfontein** [comment]
CC @Ajpantuso

**Comment by hummeltech** [comment]
This has been rebased on `main`.

**Comment by felixfontein** [comment]
I assume this isn't a WIP anymore?

**Comment by hummeltech** [comment]
Yes, I suppose not really. :+1:

File: plugins/modules/net_tools/nmcli.py [hunk 2]
```diff
@@ -1343,6 +1368,38 @@ def main():
...
+                if '802-11-wireless.%s' % name not in available_options:
+                    nmcli.module.warn(
+                        "The option 'wifi.%s' is invalid/unsupported, disregarding. Valid options are ['%s']" % (
+                            name,
+                            "', '".join(available_options).replace('802-11-wireless.', '')
+                        )
+                    )
+                    unsupported_options.append(name)
+            for unsupported_option in unsupported_options:
+                del nmcli.wifi[unsupported_option]
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:1388]: Why not parameterize with a function so the logic can be reused for `wifi_sec` instead of duplicated?

File: plugins/modules/net_tools/nmcli.py [hunk 3]
```diff
@@ -1173,6 +1170,34 @@ def show_connection(self):
...
+    def get_available_options(self, return_type, set_option=None):
+        options = []
+
+        if return_type == '802-11-wireless-security' and set_option is None:
+            set_option = 'psk'
+
+        if set_option:
+            commands = ['set %s.%s %s' % (return_type, set_option, set_option)]
+        else:
+            commands = []
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:1182]:
```suggestion
    def get_available_options(self, return_type, set_option=None):
        options = []

        if return_type == '802-11-wireless-security':
            set_option = set_option or 'psk'

        if set_option:
            commands = ['set %s.%s %s' % (return_type, set_option, set_option)]
        else:
            commands = []
```

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
From a user standpoint I like that the checks are performed before modification, but I'd prefer that there also be an option to select whether the task should fail if there are invalid options present. And I think failure should probably be the default behavior. Also updating the `wifi/wifi_sec` docs with more details on which settings are available for which versions of `nmcli` would be helpful.

File: plugins/modules/net_tools/nmcli.py (hunk 2 above)
Comment (hummeltech) [reply plugins/modules/net_tools/nmcli.py]: Alright I think that should be better, how does it look now?

**Comment by hummeltech** [comment]
I added all of the current options to the docs, unfortunately I can't do something similar with the `argument_spec` due to the hyphenations:
```diff
diff --git a/plugins/modules/net_tools/nmcli.py b/plugins/modules/net_tools/nmcli.py
index 3a2f5a12..d8a04530 100644
--- a/plugins/modules/net_tools/nmcli.py
+++ b/plugins/modules/net_tools/nmcli.py
@@ -1613,7 +1613,28 @@ def main():
             # 802-11-wireless* specific vars
             ssid=dict(type='str'),
             wifi=dict(type='dict'),
-            wifi_sec=dict(type='dict', no_log=True),
+            wifi_sec=dict(type='dict', options=dict(
+                'auth-alg'=dict(type='str'),
+                'fils'=dict(type='str'),
+                'group'=dict(type='str'),
+                'key-mgmt'=dict(type='str'),
+                'leap-password-flags'=dict(type='str'),
+                'leap-password'=dict(type='str'),
+                'leap-username'=dict(type='str'),
+                'pairwise'=dict(type='str'),
+                'pmf'=dict(type='str'),
+                'proto'=dict(type='str'),
+                'psk-flags'=dict(type='str'),
+                'psk'=dict(type='str', no_log=True),
+                'wep-key-flags'=dict(type='str'),
+                'wep-key-type'=dict(type='str'),
+                'wep-key0'=dict(type='str', no_log=True),
+                'wep-key1'=dict(type='str', no_log=True),
+                'wep-key2'=dict(type='str', no_log=True),
+                'wep-key3'=dict(type='str', no_log=True),
+                'wep-tx-keyidx'=dict(type='str'),
+                'wps-method'=dict(type='str'),
+            )),
         ),
         mutually_exclusive=[['never_default4', 'gw4']],
         required_if=[("type", "wifi", [("ssid")])],
```

File: plugins/modules/net_tools/nmcli.py [hunk 4]
```diff
@@ -345,11 +455,140 @@
...
+                default: 1
+       version_added: 3.5.0
+    ignore_unsupported_suboptions:
+       description:
+            - Ignore suboptions which are invalid or unsupported by the version of NetworkManager/nmcli
+              installed on the host.
+            - Only 'wifi' and 'wifi_sec' options are currently affected.
+       type: bool
+       default: no
        version_added: 3.5.0
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:592]:
```suggestion
       version_added: 3.6.0
```
3.5.0 has already been released.

**Comment by felixfontein** [comment]
Well, you can use `{'key-name': value}` instead of `dict(key_name=value)` to add hypens.

**BUT** you must not add `options=...` for existing options (that already exist in a release), because then Ansible will validate these options and only allow the ones specified in argument_spec to be present. This would be a breaking change.

File: plugins/modules/net_tools/nmcli.py [hunk 5]
```diff
@@ -345,12 +455,141 @@
...
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_DEFAULT (0x1) (to use global settings) and NM_SETTING_WIRELESS_WAKE_ON_WLAN_IGNORE (0x8000) (to
+                             disable management of Wake-on-LAN in NetworkManager).
+                type: int
+                default: 1
        version_added: 3.5.0
+    ignore_unsupported_suboptions:
+       description:
+            - Ignore suboptions which are invalid or unsupported by the version of NetworkManager/nmcli
+              installed on the host.
+            - Only 'wifi' and 'wifi_sec' options are currently affected.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:589]:
```suggestion
            - Only I(wifi) and I(wifi_sec) options are currently affected.
```

File: plugins/modules/net_tools/nmcli.py [hunk 6]
```diff
@@ -345,12 +455,141 @@
...
+                type: int
+                default: 0
+            wake-on-wlan:
+                description: The NMSettingWirelessWakeOnWLan options to enable. Not all devices support all options. May be any combination of
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_ANY (0x2), NM_SETTING_WIRELESS_WAKE_ON_WLAN_DISCONNECT (0x4),
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_MAGIC (0x8), NM_SETTING_WIRELESS_WAKE_ON_WLAN_GTK_REKEY_FAILURE (0x10),
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_EAP_IDENTITY_REQUEST (0x20), NM_SETTING_WIRELESS_WAKE_ON_WLAN_4WAY_HANDSHAKE (0x40),
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_RFKILL_RELEASE (0x80), NM_SETTING_WIRELESS_WAKE_ON_WLAN_TCP (0x100) or the special values
+                             NM_SETTING_WIRELESS_WAKE_ON_WLAN_DEFAULT (0x1) (to use global settings) and NM_SETTING_WIRELESS_WAKE_ON_WLAN_IGNORE (0x8000) (to
+                             disable management of Wake-on-LAN in NetworkManager).
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:581]:
```suggestion
                description: The NMSettingWirelessWakeOnWLan options to enable. Not all devices support all options. May be any combination of
                             C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_ANY) (0x2), C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_DISCONNECT) (0x4),
                             C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_MAGIC) (0x8), C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_GTK_REKEY_FAILURE) (0x10),
                             C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_EAP_IDENTITY_REQUEST) (0x20), C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_4WAY_HANDSHAKE) (0x40),
                             C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_RFKILL_RELEASE) (0x80), C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_TCP) (0x100) or the special values
                             C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_DEFAULT) (0x1) (to use global settings) and C(NM_SETTING_WIRELESS_WAKE_ON_WLAN_IGNORE) (0x8000) (to
                             disable management of Wake-on-LAN in NetworkManager).
```
I think the description should mention that users shouldn't list these constants, but add their values together and provide the resulting value here.

File: plugins/modules/net_tools/nmcli.py [hunk 7]
```diff
@@ -345,12 +455,141 @@
...
+                type: int
+                default: 0
+                choices: [ 0, 1, 2, 3 ]
+            rate:
+                description: If non-zero, directs the device to only use the specified bitrate for communication with the access point. Units are in Kb/s, ie
+                             5500 = 5.5 Mbit/s. This property is highly driver dependent and not all devices support setting a static bitrate.
+                type: int
+                default: 0
+            seen-bssids:
+                description: A list of BSSIDs (each BSSID formatted as a MAC address like "00:11:22:33:44:55") that have been detected as part of the Wi-Fi
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:564]:
```suggestion
                description: A list of BSSIDs (each BSSID formatted as a MAC address like C(00:11:22:33:44:55)) that have been detected as part of the Wi-Fi
```

File: plugins/modules/net_tools/nmcli.py [hunk 8]
```diff
@@ -345,12 +455,141 @@
...
+                default: 0
+            powersave:
+                description: One of NM_SETTING_WIRELESS_POWERSAVE_DISABLE (2) (disable Wi-Fi power saving), NM_SETTING_WIRELESS_POWERSAVE_ENABLE (3) (enable
+                             Wi-Fi power saving), NM_SETTING_WIRELESS_POWERSAVE_IGNORE (1) (don't touch currently configure setting) or
+                             NM_SETTING_WIRELESS_POWERSAVE_DEFAULT (0) (use the globally configured value). All other values are reserved.
+                type: int
+                default: 0
+                choices: [ 0, 1, 2, 3 ]
+            rate:
+                description: If non-zero, directs the device to only use the specified bitrate for communication with the access point. Units are in Kb/s, ie
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:559]:
```suggestion
                description: If non-zero, directs the device to only use the specified bitrate for communication with the access point. Units are in Kb/s, so for example
```
Please don't use Latin abbreviations.

File: plugins/modules/net_tools/nmcli.py [hunk 9]
```diff
@@ -345,12 +455,141 @@
...
+                choices: [ infrastructure, mesh, adhoc, ap ]
+                default: infrastructure
+            mtu:
+                description: If non-zero, only transmit packets of the specified size or smaller, breaking larger packets up into multiple Ethernet frames.
+                type: int
+                default: 0
+            powersave:
+                description: One of NM_SETTING_WIRELESS_POWERSAVE_DISABLE (2) (disable Wi-Fi power saving), NM_SETTING_WIRELESS_POWERSAVE_ENABLE (3) (enable
+                             Wi-Fi power saving), NM_SETTING_WIRELESS_POWERSAVE_IGNORE (1) (don't touch currently configure setting) or
+                             NM_SETTING_WIRELESS_POWERSAVE_DEFAULT (0) (use the globally configured value). All other values are reserved.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:554]:
```suggestion
                description: One of C(NM_SETTING_WIRELESS_POWERSAVE_DISABLE) (2) (disable Wi-Fi power saving), C(NM_SETTING_WIRELESS_POWERSAVE_ENABLE) (3) (enable
                             Wi-Fi power saving), C(NM_SETTING_WIRELESS_POWERSAVE_IGNORE) (1) (don't touch currently configure setting) or
                             C(NM_SETTING_WIRELESS_POWERSAVE_DEFAULT) (0) (use the globally configured value). All other values are reserved.
```

File: plugins/modules/net_tools/nmcli.py [hunk 10]
```diff
@@ -345,12 +455,141 @@
...
+                             NM_SETTING_MAC_RANDOMIZATION_ALWAYS (2) (always randomize the MAC address). This property is deprecated for 'cloned-mac-address'.
+                type: int
+                default: 0
+                choices: [ 0, 1, 2 ]
+            mac-address:
+                description: If specified, this connection will only apply to the Wi-Fi device whose permanent MAC address matches. This property does not
+                             change the MAC address of the device (i.e. MAC spoofing).
+                type: str
+            mode:
+                description: Wi-Fi network mode; one of "infrastructure", "mesh", "adhoc" or "ap". If blank, infrastructure is assumed.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:543]:
```suggestion
                description: Wi-Fi network mode. If blank, infrastructure is assumed.
```
The choices are listed next to the description, there's no need to repeat them.

File: plugins/modules/net_tools/nmcli.py [hunk 11]
```diff
@@ -345,12 +455,141 @@
...
+            mac-address-randomization:
+                description: One of NM_SETTING_MAC_RANDOMIZATION_DEFAULT (0) (never randomize unless the user has set a global default to randomize and the
+                             supplicant supports randomization), NM_SETTING_MAC_RANDOMIZATION_NEVER (1) (never randomize the MAC address), or
+                             NM_SETTING_MAC_RANDOMIZATION_ALWAYS (2) (always randomize the MAC address). This property is deprecated for 'cloned-mac-address'.
+                type: int
+                default: 0
+                choices: [ 0, 1, 2 ]
+            mac-address:
+                description: If specified, this connection will only apply to the Wi-Fi device whose permanent MAC address matches. This property does not
+                             change the MAC address of the device (i.e. MAC spoofing).
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:540]:
```suggestion
                             change the MAC address of the device (for example for MAC spoofing).
```

File: plugins/modules/net_tools/nmcli.py [hunk 12]
```diff
@@ -345,12 +455,141 @@
...
+                default: false
+            mac-address-blacklist:
+                description: A list of permanent MAC addresses of Wi-Fi devices to which this connection should never apply. Each MAC address should be given
+                             in the standard hex-digits-and-colons notation (eg "00:11:22:33:44:55").
+                type: list
+                elements: str
+            mac-address-randomization:
+                description: One of NM_SETTING_MAC_RANDOMIZATION_DEFAULT (0) (never randomize unless the user has set a global default to randomize and the
+                             supplicant supports randomization), NM_SETTING_MAC_RANDOMIZATION_NEVER (1) (never randomize the MAC address), or
+                             NM_SETTING_MAC_RANDOMIZATION_ALWAYS (2) (always randomize the MAC address). This property is deprecated for 'cloned-mac-address'.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:534]:
```suggestion
                description: One of C(NM_SETTING_MAC_RANDOMIZATION_DEFAULT) (0) (never randomize unless the user has set a global default to randomize and the
                             supplicant supports randomization), C(NM_SETTING_MAC_RANDOMIZATION_NEVER) (1) (never randomize the MAC address), or
                             C(NM_SETTING_MAC_RANDOMIZATION_ALWAYS) (2) (always randomize the MAC address). This property is deprecated for 'cloned-mac-address'.
```

File: plugins/modules/net_tools/nmcli.py [hunk 13]
```diff
@@ -345,12 +455,141 @@
...
+                             mode. In infrastructure mode, various workarounds are used for a more reliable discovery of hidden networks, such as
+                             probe-scanning the SSID. However, these workarounds expose inherent insecurities with hidden SSID networks, and thus hidden SSID
+                             networks should be used with caution. In AP mode, the created network does not broadcast its SSID. Note that marking the network
+                             as hidden may be a privacy issue for you (in infrastructure mode) or client stations (in AP mode), as the explicit probe-scans are
+                             distinctly recognizable on the air.
+                type: bool
+                default: false
+            mac-address-blacklist:
+                description: A list of permanent MAC addresses of Wi-Fi devices to which this connection should never apply. Each MAC address should be given
+                             in the standard hex-digits-and-colons notation (eg "00:11:22:33:44:55").
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:528]:
```suggestion
                             in the standard hex-digits-and-colons notation (for example, C(00:11:22:33:44:55)).
```

**Comment by hummeltech** [comment]
Thanks for the tip, I didn't see mention of that syntax being supported in Ansible's argument spec spec. Yes, that's true, however if all past and present sub-options were available in the `argument_spec` then I don't believe there would be any breaking changes. The only change would be that future sub-options (properties) would need to be added in order to be supported.

File: plugins/modules/net_tools/nmcli.py [hunk 14]
```diff
@@ -345,12 +455,141 @@
...
+                             "02:00:00:00:00:00 00:00:00:00:00:00 02:00:00:00:00:00" will create a fully scrambled MAC address, randomly locally or globally
+                             administered.
+                type: str
+            hidden:
+                description: If TRUE, indicates that the network is a non-broadcasting network that hides its SSID. This works both in infrastructure and AP
+                             mode. In infrastructure mode, various workarounds are used for a more reliable discovery of hidden networks, such as
+                             probe-scanning the SSID. However, these workarounds expose inherent insecurities with hidden SSID networks, and thus hidden SSID
+                             networks should be used with caution. In AP mode, the created network does not broadcast its SSID. Note that marking the network
+                             as hidden may be a privacy issue for you (in infrastructure mode) or client stations (in AP mode), as the explicit probe-scans are
+                             distinctly recognizable on the air.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:523]: 1. Boolean values are `true` and `false`.
2. Break up too long paragraphs into multiple paragraphs.
```suggestion
                description:
                  - If C(true), indicates that the network is a non-broadcasting network that hides its SSID. This works both in infrastructure and AP
                    mode.
                  - In infrastructure mode, various workarounds are used for a more reliable discovery of hidden networks, such as
                    probe-scanning the SSID. However, these workarounds expose inherent insecurities with hidden SSID networks, and thus hidden SSID
                    networks should be used with caution.
                  - In AP mode, the created network does not broadcast its SSID. Note that marking the network
                    as hidden may be a privacy issue for you (in infrastructure mode) or client stations (in AP mode), as the explicit probe-scans are
                    distinctly recognizable on the air.
```

**Review by felixfontein — COMMENTED** [review]
Please use correct formatting for the descriptions. See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation for the available options.

I've added a few comments, thre are a lot more places where similar ones apply

**Comment by felixfontein** [comment]
> Thanks for the tip, I didn't see mention of that syntax being supported in Ansible's argument spec spec. Yes, that's true, however if all past and present sub-options were available in the `argument_spec` then I don't believe there would be any breaking changes.

It would still be a breaking change since the old version works with any new nmcli parameter out of the box, while the new version would not. Users could be installing the old version of the collection also in a year, instead of using a later version, and then upgrade. For them this could break something.

**Comment by hummeltech** [comment]
Perhaps that is something that can be done for `v4.0.0` then?

**Comment by felixfontein** [comment]
> Perhaps that is something that can be done for `v4.0.0` then?

Yes, that would be ok from my POV.

File: plugins/modules/net_tools/nmcli.py [hunk 15]
```diff
@@ -332,11 +332,141 @@
...
               U(https://networkmanager.dev/docs/api/latest/settings-802-11-wireless-security.html).'
             - 'For instance to use common WPA-PSK auth with a password:
               C({key-mgmt: wpa-psk, psk: my_password}).'
        type: dict
+       suboptions:
+            auth-alg:
+                description:
+                    - When WEP is used (ie, I(key-mgmt) = C(none) or C(ieee8021x)) indicate the 802.11 authentication algorithm required by the AP here.
+                    - One of C(open) for Open System, C(shared) for Shared Key, or C(leap) for Cisco LEAP.
+                    - When using Cisco LEAP (ie, I(key-mgmt) = C(ieee8021x) and I(auth-alg) = C(leap)) the I(leap-username) and I(leap-password) properties
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:347]:
```suggestion
                    - When using Cisco LEAP (that is, if I(key-mgmt=ieee8021x) and I(auth-alg=leap)) the I(leap-username) and I(leap-password) properties
```

File: plugins/modules/net_tools/nmcli.py [hunk 16]
```diff
@@ -332,11 +332,141 @@
...
+            - Note the list of suboption attributes may vary depending on which version of NetworkManager/nmcli is installed on the host.
+            - 'An up-to-date list of supported attributes can be found here:
               U(https://networkmanager.dev/docs/api/latest/settings-802-11-wireless-security.html).'
             - 'For instance to use common WPA-PSK auth with a password:
               C({key-mgmt: wpa-psk, psk: my_password}).'
        type: dict
+       suboptions:
+            auth-alg:
+                description:
+                    - When WEP is used (ie, I(key-mgmt) = C(none) or C(ieee8021x)) indicate the 802.11 authentication algorithm required by the AP here.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:345]:
```suggestion
                    - When WEP is used (that is, I(key-mgmt=none) or C(key-mgmt=ieee8021x)) indicate the 802.11 authentication algorithm required by the AP here.
```

File: plugins/modules/net_tools/nmcli.py [hunk 17]
```diff
@@ -332,11 +332,141 @@
...
+                      + WPA3 personal), C(sae) (WPA3 personal only), C(wpa-eap) (WPA2 + WPA3 enterprise) or C(wpa-eap-suite-b-192) (WPA3 enterprise only).
+                    - This property must be set for any Wi-Fi connection that uses security.
+                type: str
+                choices: [ none, ieee8021x, owe, wpa-psk, sae, wpa-eap, wpa-eap-suite-b-192 ]
+            leap-password-flags:
+                description: Flags indicating how to handle the I(leap-password) property.
+                type: list
+                elements: int
+            leap-password:
+                description: The login password for legacy LEAP connections (ie, I(key-mgmt) = C(ieee8021x) and I(auth-alg) = C(leap)).
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:381]:
```suggestion
                description: The login password for legacy LEAP connections (ie, I(key-mgmt=ieee8021x) and I(auth-alg=leap)).
```

File: plugins/modules/net_tools/nmcli.py [hunk 18]
```diff
@@ -332,11 +332,141 @@
...
+                type: str
+            leap-username:
+                description: The login username for legacy LEAP connections (ie, I(key-mgmt) = C(ieee8021x) and I(auth-alg) = C(leap)).
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:384]:
```suggestion
                description: The login username for legacy LEAP connections (ie, I(key-mgmt=ieee8021x) and I(auth-alg=leap)).
```

File: plugins/modules/net_tools/nmcli.py [hunk 19]
```diff
@@ -332,11 +332,141 @@
...
+                    - One of C(0) (use global default value), C(1) (disable PMF), C(2) (enable PMF if the supplicant and the access point support it) or C(3)
+                      (enable PMF and fail if not supported).
+                    - When set to C(0) and no global default is set, PMF will be optionally enabled.
+                type: int
+                choices: [ 0, 1, 2, 3 ]
+                default: 0
+            proto:
+                description:
+                    - List of strings specifying the allowed WPA protocol versions to use.
+                    - Each element may be one C(wpa) (allow WPA) or C(rsn) (allow WPA2/RSN).
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:406]:
```suggestion
                    - Each element may be C(wpa) (allow WPA) or C(rsn) (allow WPA2/RSN).
```

File: plugins/modules/net_tools/nmcli.py [hunk 20]
```diff
@@ -332,11 +332,141 @@
...
+                    - See the I(wep-key-type) property for a description of how this key is interpreted.
+                type: str
+            wep-key3:
+                description:
+                    - Index 3 WEP key. This WEP index is not used by most networks.
+                    - See the I(wep-key-type) property for a description of how this key is interpreted.
+                type: str
+            wep-tx-keyidx:
+                description:
+                    - When static WEP is used (ie, I(key-mgmt) = C(none)) and a non-default WEP key index is used by the AP, put that WEP key index here.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:456]:
```suggestion
                    - When static WEP is used (that is, I(key-mgmtnone)) and a non-default WEP key index is used by the AP, put that WEP key index here.
```

File: plugins/modules/net_tools/nmcli.py [hunk 21]
```diff
@@ -332,11 +332,141 @@
...
+                    - Valid values are C(0) (default key) through C(3).
+                    - Note that some consumer access points (like the Linksys WRT54G) number the keys C(1) - C(4).
+                type: int
+                choices: [ 0, 1, 2, 3 ]
+                default: 0
+            wps-method:
+                description:
+                    - Flags indicating which mode of WPS is to be used if any.
+                    - There's little point in changing the default setting as NetworkManager will automatically determine whether it's feasible to start WPS
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:465]:
```suggestion
                    - There is little point in changing the default setting as NetworkManager will automatically determine whether it is feasible to start WPS
```

File: plugins/modules/net_tools/nmcli.py [hunk 22]
```diff
@@ -332,11 +332,141 @@
...
+                elements: str
+                choices: [ wpa, rsn ]
+            psk-flags:
+                description: Flags indicating how to handle the I(psk) property.
+                type: list
+                elements: int
+            psk:
+                description:
+                    - Pre-Shared-Key for WPA networks.
+                    - For WPA-PSK, it's either an ASCII passphrase of 8 to 63 characters that is (as specified in the 802.11i standard) hashed to derive the
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:418]:
```suggestion
                    - For WPA-PSK, it is either an ASCII passphrase of 8 to 63 characters that is (as specified in the 802.11i standard) hashed to derive the
```

**Review by felixfontein — APPROVED** [review]

File: plugins/modules/net_tools/nmcli.py [hunk 23]
```diff
@@ -1173,6 +1451,62 @@ def show_connection(self):
...
+    def get_supported_properties(self, setting, set_property=None, set_value='FAKEVALUE'):
+        properties = []
+
+        if setting == '802-11-wireless-security':
+            set_property = set_property or 'psk'
+
+        if set_property:
+            commands = ['set %s.%s %s' % (setting, set_property, set_value)]
+        else:
+            commands = []
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:1463]:
```suggestion
    def get_supported_properties(self, setting):
        properties = []

        if setting == '802-11-wireless-security':
            set_property = 'psk'
            set_value='FAKEVALUE'
            commands = ['set %s.%s %s' % (setting, set_property, set_value)]
        else:
            commands = []
```
I think this is over generalized at the moment. It can be refactored later if needed, but this is more clear as a special case for `'802-11-wireless-security'` IMO.

File: plugins/modules/net_tools/nmcli.py [hunk 24]
```diff
@@ -1173,6 +1451,62 @@ def show_connection(self):
...
+        for line in out.splitlines():
+            prefix = '%s.' % setting
+            if (line.startswith(prefix)):
+                pair = line.split(':', 1)
+                property = pair[0].strip().replace(prefix, '')
+                properties.append(property)
+
+        return properties
+
+    def check_supported_properties(self, setting):
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:1483]: Can this function be broken down further? At least the deletion of unsupported properties should be separated since that is outside the scope of "checking".

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
Just a few minor items to consider.

File: plugins/modules/net_tools/nmcli.py (hunk 23 above)
Comment (hummeltech) [reply plugins/modules/net_tools/nmcli.py:1463]: Yes, it will very likely apply to other `properties` (if the use of this method is expanded), but the `property` currently using this exception is indeed `802-11-wireless-security`.

File: plugins/modules/net_tools/nmcli.py (hunk 24 above)
Comment (hummeltech) [reply plugins/modules/net_tools/nmcli.py:1483]: Sure, I can mess around with that.

Comment (hummeltech) [reply plugins/modules/net_tools/nmcli.py]: Alright, I moved the `unsupported` property `deletion` into the `main` function, hopefully that at least partially satisfies your concern.

**Review by Ajpantuso — APPROVED** [review]
LGTM, thanks!

**Comment by felixfontein** [comment]
@hummeltech thanks for implementing this!
@Ajpantuso thanks for reviewing!
